fc/code_generation.py

```diff
@@ -59,6 +60,67 @@ def load_template(*name):
     return env.get_template(path)
 
 
+class DataInterpolation(Quantity):
+    """Represents linear interpolation on data loaded from file, for code generation."""
+    _next_id = 0  # Used to generate unique table IDs in code
+
+    # Sympy annoyingly overwrites __new__
+    def __new__(cls, name, data, index_variable, *args, **kwargs):
+        print('interp__new__', name)
+        obj = super().__new__(cls, name, real=True)
+
+        # Record a unique ID for this table
+        obj._id = cls._next_id
+        cls._next_id += 1
+
+        # Ensure we depend on the index variable in the model's graph
+        obj._args = [index_variable]
+
+        return obj
+
+    def __init__(self, name, data, index_variable, units):
+        """Create a new interpolation construct.
+
+        :param name: the data file base name
+        :param data: 2d numpy array containing the column-wise data
+        :param index_variable: the :class:`cellmlmanip.model.Variable` used to index the data
+        :param units: the units of the interpolated values (a :class:`~cellmlmanip.units.UnitStore.Unit`)
+        """
+        print('interp__init__', name, index_variable, units)
+        super().__init__(name, units)
+        self.data = data[1]
+        self.index_variable = index_variable
+        self.initial_index = '%.17g' % data[0, 0]
+        self.final_index = '%.17g' % data[0, -1]
+
+        self.table_name = '_data_table_' + str(self._id)
+        self.step_inverse = '%.17g' % (1.0 / (data[0, 1] - data[0, 0]))
+
+        self.index_name = None
+
+    def _eval_evalf(self, prec):
+        """We don't want the model's graph to try replacing this with a number!"""
+        return self
+
+    def set_index_name(self, variable_name_generator):
+        """Set the name used for our index variable in code.
+
+        :param variable_name_generator: function turning our index variable into its name in the code
+        """
+        self.index_name = variable_name_generator(self.index_variable)
+
+    @property
+    def lookup_call(self):
+        """The code for a function call looking up this interpolation table."""
+        assert self.index_name is not None
```

Comment:
`RuntimeError` ?
Looks like misuse of the interface rather than a sanity check?

Reply:
It's only code in the same module that uses this bit of the interface though, so felt more like an internal consistency check, and not worth adding a test of the error case for.